File: Leetcode2075. Decode the slanted ciphertext/Solution.py
class Solution:
    def decodeCiphertext(self, encodedText: str, rows: int) -> str:
        if rows == 1:
            return encodedText

        n = len(encodedText)
        cols = n // rows
        res = []

        for c in range(cols):
            r, j = 0, c
            while r < rows and j < cols:
                res.append(encodedText[r * cols + j])
                r += 1
                j += 1

        return "".join(res).rstrip()



# runtime - 2ms

class Solution:
    def decodeCiphertext(self, encoded_text: str, rows: int) -> str:
        if rows == 1:
            return encoded_text

        N = len(encoded_text)
        cols = N // rows
        i, j, k = 0, 0, 0
        original_text = []

        while k < N:
            original_text.append(encoded_text[k])
            i += 1
            if i == rows:
                i = 0
                j += 1
            k = i*(cols + 1) + j

        return ''.join(original_text).rstrip()



# runtime - 8ms

class Solution:
    def decodeCiphertext(self, encoded_text: str, rows: int) -> str:
        if rows == 1:
            return encoded_text

        N = len(encoded_text)
        cols = N // rows
        i, j, k = 0, 0, 0
        original_text = []

        while k < N:
            original_text.append(encoded_text[k])
            i += 1
            if i == rows:
                i = 0
                j += 1
            k = i*(cols + 1) + j

        return ''.join(original_text).rstrip()



# runtime - 13ms

class Solution:
    def decodeCiphertext(self, encoded_text: str, rows: int) -> str:
        if rows == 1:
            return encoded_text

        N = len(encoded_text)
        cols = N // rows
        matrix = [encoded_text[i * cols:(i + 1) * cols] for i in range(rows)]
        original = []

        for c in range(cols):
            r, cc = 0, c
            while r < rows and cc < cols:
                original.append(matrix[r][cc])
                r += 1
                cc += 1

        return ''.join(original).rstrip()



# runtime - 19ms

class Solution:
    def decodeCiphertext(self, encoded_text: str, rows: int) -> str:
        if rows == 1:
            return encoded_text

        n = len(encoded_text)
        cols = n // rows
        grid = [encoded_text[r * cols:(r + 1) * cols] for r in range(rows)]

        res = []
        for c in range(cols):
            r, cc = 0, c
            while r < rows and cc < cols:
                res.append(grid[r][cc])
                r += 1
                cc += 1

        return ''.join(res).rstrip()

File: Leetcode2075. Decode the slanted ciphertext/Solution.java
class Solution {
    public String decodeCiphertext(String encodedText, int rows) {
        if (rows == 1)
            return encodedText;

        int n = encodedText.length();
        int cols = n / rows;

        StringBuilder res = new StringBuilder(n);

        for (int c = 0; c < cols; c++) {
            int r = 0, j = c;
            while (r < rows && j < cols) {
                res.append(encodedText.charAt(r * cols + j));
                r++;
                j++;
            }
        }

        int end = res.length() - 1;
        while (end >= 0 && res.charAt(end) == ' ') {
            end--;
        }

        return res.substring(0, end + 1);
    }
}


// runtime - 9ms
class Solution {
    public String decodeCiphertext(String encodedText, int rows) {
        if (rows == 1) return encodedText;
        int encodedSize = encodedText.length();
        int columns = encodedSize/rows;
        char[][] matrix = new char[rows][columns];

        int c = 0;
        for (int i = 0; i < rows; i++) {
            if (c == encodedSize) break;
            for (int j = 0; j < columns; j++) {
                if (c == encodedSize) break;
                matrix[i][j] = encodedText.charAt(c);
                c++;
            }
        }

        StringBuilder originalText = new StringBuilder();
        
        int row = 0, column = 0;
        int columnCounter = 0;
        while (column < columns) {
            char character = matrix[row][column];
            if (character == '\0') break;
            originalText.append(character);
            row++;
            column++;
            if (row == rows) {
                row = 0;
                columnCounter++;
                column = columnCounter;
            };

        }
        int limit = originalText.length();
        for (int i = originalText.length() - 1; i >= 0; i--) {
            if (originalText.charAt(i) == ' ') {
                limit--;
            } else {
                break;
            }
        }
        originalText.setLength(limit);
        return originalText.toString();
    }
}



// runtime - 17ms
class Solution {
    public String decodeCiphertext(String encodedText, int rows) {
        if (rows == 1) return encodedText;

        int n = encodedText.length();
        int cols = n / rows;

        StringBuilder result = new StringBuilder();

        // start from each column of first row
        for (int startCol = 0; startCol < cols; startCol++) {

            int r = 0;
            int c = startCol;

            // traverse diagonal
            while (r < rows && c < cols) {
                result.append(encodedText.charAt(r * cols + c));
                r++;
                c++;
            }
        }

        // remove trailing spaces
        int end = result.length() - 1;
        while (end >= 0 && result.charAt(end) == ' ') {
            end--;
        }

        return result.substring(0, end + 1);
    }
}


// runtime - 21ms
class Solution {
    public String decodeCiphertext(String encodedText, int rows) {
        if(rows == 1)
            return encodedText;

        int len = encodedText.length();

        if(len == 0)
            return encodedText;

        int cols = (int) Math.ceil( (double) len/rows);

        char[][] matrix = new char[rows][cols];
        char[] encodedArr = encodedText.toCharArray();
        int e=0;

        for(int i=0;i<rows;i++) {
            for(int j=0;j<cols;j++) {
                matrix[i][j] = encodedArr[e++];
            }
        }

        StringBuilder sb = new StringBuilder();

        for (int startCol = 0; startCol < cols; startCol++) {
            int r = 0;
            int c = startCol;

            while (r < rows && c < cols) {
                sb.append(matrix[r][c]);
                r++;
                c++;
            }
        }

        return sb.toString().stripTrailing();
    }
}
